src/ndt/protocol_impl.hpp

```diff
@@ -0,0 +1,260 @@
+            return;
+        }
+        std::string s = ctx->buff->readn(KICKOFF_MESSAGE_SIZE);
+        if (s != KICKOFF_MESSAGE) {
+            callback(InvalidKickoffMessageError());
+            return;
+        }
+        ctx->logger->info("Got legacy KICKOFF message (ignored)");
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(messages, read_msg)>
+void wait_in_queue_impl(Var<Context> ctx, Callback<Error> callback) {
+    ctx->logger->debug("ndt: wait in queue ...");
+    messages_read_msg(ctx, [=](Error err, uint8_t type, std::string s) {
+        ctx->logger->debug("ndt: wait in queue ... %d", (int)err);
+        if (err) {
+            callback(ReadingSrvQueueMessageError(err));
+            return;
+        }
+        if (type != SRV_QUEUE) {
+            callback(NotSrvQueueMessageError());
+            return;
+        }
+        ErrorOr<unsigned> wait_time = lexical_cast_noexcept<unsigned>(s);
+        if (!wait_time) {
+            callback(InvalidSrvQueueMessageError(wait_time.as_error()));
+            return;
+        }
+        ctx->logger->info("Wait time before test starts: %d", *wait_time);
+        // XXX Simplified implementation
+        if (*wait_time > 0) {
+            callback(UnhandledSrvQueueMessageError());
+            return;
+        }
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(messages, read_msg)>
+void recv_version_impl(Var<Context> ctx, Callback<Error> callback) {
+    ctx->logger->debug("ndt: recv server version ...");
+    messages_read_msg(ctx, [=](Error err, uint8_t type, std::string s) {
+        ctx->logger->debug("ndt: recv server version ... %d", (int)err);
+        if (err) {
+            callback(ReadingServerVersionMessageError(err));
+            return;
+        }
+        if (type != MSG_LOGIN) {
+            callback(NotServerVersionMessageError());
+            return;
+        }
+        ctx->logger->info("Got server version: %s", s.c_str());
+        // TODO: validate the server version?
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(messages, read_msg)>
+void recv_tests_id_impl(Var<Context> ctx, Callback<Error> callback) {
+    ctx->logger->debug("ndt: recv tests ID ...");
+    messages_read_msg(ctx, [=](Error err, uint8_t type, std::string s) {
+        ctx->logger->debug("ndt: recv tests ID ... %d", (int)err);
+        if (err) {
+            callback(ReadingTestsIdMessageError(err));
+            return;
+        }
+        if (type != MSG_LOGIN) {
+            callback(NotTestsIdMessageError());
+            return;
+        }
+        ctx->logger->info("Authorized tests: %s", s.c_str());
+        ctx->granted_suite = split(s);
+        callback(NoError());
+    });
+}
+
+template <MK_MOCK_NAMESPACE(test_c2s, run), MK_MOCK_NAMESPACE(test_meta, run),
+          MK_MOCK_NAMESPACE(test_s2c, run)>
+void run_tests_impl(Var<Context> ctx, Callback<Error> callback) {
```

Comment:
I am thinking that it may be useful for a user of the API to be able to run specific tests. My understanding of how this call works is that you look at the tests that are supported by the NDT server and run through them in the order by which they are returned from the server.

Wouldn't it be better instead to have the order, and the types of tests to run, be defined by a user of the `NdtTest` DSL?

Perhaps having a call to set the types of test to run.